File: android/app/src/main/java/com/vibetuned/midisink/ControlStripView.kt
package com.vibetuned.midisink

import android.app.AlertDialog
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.os.SystemClock
import android.text.InputType
import android.view.MotionEvent
import android.view.View
import android.widget.EditText
import android.widget.Toast
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/**
 * Performance control strip (PHASE4_SPEC §8, DECISIONS_3 #30/#31): a compact
 * FLOATING palette over the full-canvas lattice — never a docked band
 * (displacing the lattice broke drop-under-finger on device). Wheels and a
 * button built from the joystick primitive: a touch anchors its origin and
 * the §3.2 soft knee shapes Δy (via hostmpe, one implementation). All VALUE
 * state lives in hostmpe_strip_t on the MIDI thread; this view keeps display
 * mirrors only. Every message the strip causes goes out on the master
 * channel — the widget engines enforce that, unit-tested headlessly.
 *
 * Touch hygiene learned on iOS (#31): every widget touch is tracked in the
 * grab table (sustain included) so a release resolves by its grab record,
 * and the long-press CC editor exists ONLY for the two assignable wheels —
 * nothing may ever cancel a held sustain pad.
 */
class ControlStripView(context: Context) : View(context) {

    enum class Widget { PITCH, MOD, ASSIGN_A, ASSIGN_B, SUSTAIN }

    private val density = resources.displayMetrics.density
    // Travel bound for the wheel joysticks, in dp (≙ iOS points): the §3.2
    // knee needs Δ and r_max in ONE metric; the strip is a fixed-size
    // palette, not the lattice.
    private val wheelTravelDp = 60f
    // Full-travel latch sweep in CC units per grab (slow drags accumulate).
    private val latchSweep = 64f

    private class Grab(val widget: Widget, val originY: Float, val originX: Float) {
        var lastShaped = 0f
        var moved = false
        var longPress: Runnable? = null
    }
    private val grabs = HashMap<Int, Grab>()

    // Display mirrors (the engine on the MIDI thread is the truth).
    private var mirrorPitch = 0f
    private val mirrorLatch = floatArrayOf(0f, 0f, 0f)
    private var mirrorSustain = false
    private var mirrorToggleMode = false
    private val mirrorCCs = intArrayOf(1, 23, 24)
    private var ramping = false
    private val effOut = FloatArray(2)
    private val stateOut = FloatArray(8)

    private val paintBg = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
        color = Color.argb((0.42f * 255).toInt(), 255, 255, 255)
    }
    private val paintBorder = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        color = Color.argb((0.15f * 255).toInt(), 0, 0, 0)
        strokeWidth = 0.5f * density
    }
    private val paintSlot = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        color = Color.argb((0.25f * 255).toInt(), 0, 0, 0)
        strokeWidth = 1f * density
    }
    private val paintTrack = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
        color = Color.argb((0.12f * 255).toInt(), 0, 0, 0)
    }
    private val paintNotch = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
        color = Color.argb((0.30f * 255).toInt(), 0, 0, 0)
    }
    private val paintThumb = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
        color = Color.argb((0.55f * 255).toInt(), 0, 0, 0)
    }
    private val paintPadOn = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
        color = Color.argb((0.45f * 255).toInt(), 0, 0, 0)
    }
    private val paintPadOff = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        color = Color.argb((0.40f * 255).toInt(), 0, 0, 0)
        strokeWidth = 1f * density
    }
    private val paintText = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.argb((0.60f * 255).toInt(), 0, 0, 0)
        textSize = 11f * density
        textAlign = Paint.Align.CENTER
    }

    init {
        setWillNotDraw(false)
    }

    /** Pull the engine's latched state (mode re-entry, sustain-mode changes). */
    fun syncMirrors(toggleMode: Boolean) {
        NativeBridge.nativeStripState(stateOut)
        mirrorPitch = stateOut[0]
        for (w in 0 until 3) {
            mirrorLatch[w] = stateOut[1 + w]
            mirrorCCs[w] = stateOut[5 + w].toInt()
        }
        mirrorSustain = stateOut[4] > 0.5f
        mirrorToggleMode = toggleMode
        invalidate()
    }

    // -- geometry ------------------------------------------------------------------

    private fun slotRect(w: Widget): RectF {
        val n = Widget.values().size
        val sw = width.toFloat() / n
        val inset = 6f * density
        return RectF(w.ordinal * sw + inset, inset, (w.ordinal + 1) * sw - inset, height - inset)
    }

    private fun widgetAt(x: Float, y: Float): Widget? {
        val slack = 6f * density
        for (w in Widget.values()) {
            val r = slotRect(w)
            if (x >= r.left - slack && x <= r.right + slack && y >= r.top - slack && y <= r.bottom + slack) return w
        }
        return null
    }

    // -- touches: the joystick primitive per widget --------------------------------

    /** Clamped, knee-shaped vertical deflection for a wheel grab (up = +1). */
    private fun shaped(dyPx: Float): Float {
        NativeBridge.nativeJoystickEff(0f, dyPx / density, wheelTravelDp, effOut)
        return -effOut[1]   // screen y grows down; up is positive
    }

    override fun onTouchEvent(e: MotionEvent): Boolean {
        when (e.actionMasked) {
            MotionEvent.ACTION_DOWN, MotionEvent.ACTION_POINTER_DOWN -> {
                val i = e.actionIndex
                val x = e.getX(i)
                val y = e.getY(i)
                val w = widgetAt(x, y) ?: return true
                val id = e.getPointerId(i)
                val g = Grab(w, y, x)
                grabs[id] = g
                when (w) {
                    Widget.SUSTAIN -> {
                        mirrorSustain = if (mirrorToggleMode) !mirrorSustain else true
                        NativeBridge.nativeStripSustainDown()
                    }
                    Widget.PITCH -> ramping = false
                    Widget.ASSIGN_A, Widget.ASSIGN_B -> {
                        // The editor gesture is a hand-rolled long-press that
                        // exists ONLY for the assignable wheels (#31).
                        val lp = Runnable {
                            val cur = grabs[id]
                            if (cur === g && !g.moved) {
                                grabs.remove(id)
                                openEditor(w)
                            }
                        }
                        g.longPress = lp
                        postDelayed(lp, 500)
                    }
                    Widget.MOD -> {}
                }
            }
            MotionEvent.ACTION_MOVE -> {
                for (i in 0 until e.pointerCount) {
                    val g = grabs[e.getPointerId(i)] ?: continue
                    val y = e.getY(i)
                    if (!g.moved && (abs(y - g.originY) > 8f * density || abs(e.getX(i) - g.originX) > 8f * density)) {
                        g.moved = true
                        g.longPress?.let { removeCallbacks(it) }
                        g.longPress = null
                    }
                    val s = shaped(y - g.originY)
                    when (g.widget) {
                        Widget.PITCH -> {
                            mirrorPitch = s
                            NativeBridge.nativeStripPitchMove(s)
                        }
                        Widget.MOD, Widget.ASSIGN_A, Widget.ASSIGN_B -> {
                            // Relative accumulation (§8): only the CHANGE in
                            // shaped position moves the value — a regrasp
                            // contributes zero.
                            val wheel = g.widget.ordinal - Widget.MOD.ordinal
                            val delta = (s - g.lastShaped) * latchSweep
                            mirrorLatch[wheel] = (mirrorLatch[wheel] + delta).coerceIn(0f, 127f)
                            NativeBridge.nativeStripLatchMove(wheel, delta)
                        }
                        Widget.SUSTAIN -> {}
                    }
                    g.lastShaped = s
                }
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_POINTER_UP -> release(e.getPointerId(e.actionIndex))
            MotionEvent.ACTION_CANCEL -> {
                for (id in grabs.keys.toList()) release(id)
            }
        }
        invalidate()
        return true
    }

    private fun release(id: Int) {
        val g = grabs.remove(id) ?: return
        g.longPress?.let { removeCallbacks(it) }
        when (g.widget) {
            Widget.PITCH -> {
                NativeBridge.nativeStripPitchRelease()
                animateSpringReturn()
            }
            Widget.SUSTAIN -> {
                if (!mirrorToggleMode) mirrorSustain = false
                NativeBridge.nativeStripSustainUp()   // toggle mode: engine-side no-op
            }
            Widget.MOD, Widget.ASSIGN_A, Widget.ASSIGN_B -> {}
        }
    }

    // Visual mirror of the engine's 50 ms return ramp (the MIDI ramp itself is
    // emitted by hostmpe_strip_tick on the MIDI thread's drain).
    private fun animateSpringReturn() {
        ramping = true
        val v0 = mirrorPitch
        val t0 = SystemClock.uptimeMillis()
        val step = object : Runnable {
            override fun run() {
                if (!ramping) return
                val f = (SystemClock.uptimeMillis() - t0) / 50f
                if (f >= 1f) {
                    mirrorPitch = 0f
                    ramping = false
                } else {
                    mirrorPitch = v0 * (1f - f)
                    postOnAnimation(this)
                }
                invalidate()
            }
        }
        step.run()
    }

    // -- the assignable-wheel CC editor (§8) -----------------------------------------

    private fun openEditor(w: Widget) {
        val wheel = w.ordinal - Widget.MOD.ordinal
        val current = mirrorCCs[wheel]
        val input = EditText(context).apply {
            inputType = InputType.TYPE_CLASS_NUMBER
            setText(current.toString())
            // Typing REPLACES the current number: the field opens pre-filled,
            // and without this every keystroke inserts at the cursor (a tap
            // then "25" over "23" gives 2523, silently out of range).
            setSelectAllOnFocus(true)
            requestFocus()
        }
        AlertDialog.Builder(context)
            .setTitle("Assign wheel ${if (w == Widget.ASSIGN_A) "A" else "B"}")
            .setMessage("7-bit CC number (0–127). Currently CC $current. " +
                "Protocol CCs (1, 6, 38, 64, 98–101, 120–127) are refused.")
            .setView(input)
            .setNegativeButton("Cancel", null)
            .setPositiveButton("Assign") { _, _ ->
                val cc = input.text.toString().trim().toIntOrNull()
                if (cc == null || cc < 0 || cc > 127) {
                    // Say so — a silently ignored entry is indistinguishable
                    // from a refused one.
                    Toast.makeText(context, "Enter a CC number from 0 to 127",
                                   Toast.LENGTH_SHORT).show()
                    return@setPositiveButton
                }
                val assigned = NativeBridge.nativeStripAssign(wheel, cc)
                if (assigned >= 0) {
                    mirrorCCs[wheel] = assigned
                } else {
                    // The engine refuses the protocol CCs (§8 / DECISIONS_3
                    // #30): a strip-assigned CC 6 on the master would corrupt
                    // the DAW's RPN handshake mid-performance.
                    Toast.makeText(context, "CC $cc is a protocol CC — refused",
                                   Toast.LENGTH_SHORT).show()
                }
                invalidate()
            }
            .show()
    }

    // -- drawing -----------------------------------------------------------------------

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val corner = 12f * density
        canvas.drawRoundRect(0f, 0f, width.toFloat(), height.toFloat(), corner, corner, paintBg)
        canvas.drawRoundRect(0f, 0f, width.toFloat(), height.toFloat(), corner, corner, paintBorder)
        for (w in Widget.values()) {
            val r = slotRect(w)
            canvas.drawRoundRect(r, 8f * density, 8f * density, paintSlot)
            val label = when (w) {
                Widget.PITCH -> "Pitch"
                Widget.MOD -> "Mod"
                Widget.ASSIGN_A -> "CC ${mirrorCCs[1]}"
                Widget.ASSIGN_B -> "CC ${mirrorCCs[2]}"
                Widget.SUSTAIN -> if (mirrorToggleMode) "Sus ⇥" else "Sus"
            }
            canvas.drawText(label, r.centerX(), r.bottom - 5f * density, paintText)

            if (w == Widget.SUSTAIN) {
                val pad = RectF(r.left + r.width() * 0.22f, r.top + r.height() * 0.24f - 6f * density,
                                r.right - r.width() * 0.22f, r.bottom - r.height() * 0.24f - 6f * density)
                canvas.drawRoundRect(pad, 6f * density, 6f * density,
                                     if (mirrorSustain) paintPadOn else paintPadOff)
                continue
            }
            // Wheel track + thumb.
            val track = RectF(r.centerX() - 1.5f * density, r.top + 8f * density,
                              r.centerX() + 1.5f * density, r.bottom - 24f * density)
            canvas.drawRect(track, paintTrack)
            val frac = if (w == Widget.PITCH) {
                canvas.drawRect(r.centerX() - 8f * density, track.centerY() - 0.5f * density,
                                r.centerX() + 8f * density, track.centerY() + 0.5f * density, paintNotch)
                0.5f + 0.5f * mirrorPitch
            } else {
                mirrorLatch[w.ordinal - Widget.MOD.ordinal] / 127f
            }
            val ty = track.bottom - frac * track.height()
            canvas.drawCircle(r.centerX(), ty, 7f * density, paintThumb)
        }
    }

    @Suppress("unused")
    private fun clamp01(v: Float) = max(0f, min(1f, v))
}

File: ios/Sources/ControlStripView.swift
// Performance control strip (PROJECT_SPEC.md §8.8, Step 18): a dockable bar of
// wheels and buttons built from the existing joystick primitive — a touch
// anchors its origin and the §3.2 soft knee shapes Δy (via hostmpe, one
// implementation). All VALUE state lives in hostmpe_strip_t on the host's
// MIDI queue; this view keeps display mirrors only. Every message the strip
// causes goes out on the master channel — the widget engines enforce that,
// unit-tested headlessly.
import UIKit
import HostMPE

final class ControlStripView: UIView, UIGestureRecognizerDelegate {
    weak var host: SumiCanvasView?

    // Widget slots, left to right. rawValue orders the layout.
    enum Widget: Int, CaseIterable {
        case pitch = 0, mod, assignA, assignB, sustain
    }

    // Travel bound for the wheel joysticks, in points: the §3.2 knee needs Δ
    // and r_max in ONE metric; points are that metric here (the strip is a
    // fixed-height bar, not the lattice — canvas-height units would tie the
    // feel to the canvas size).
    private let wheelTravelPt: Float = 60
    // Full-travel latch sweep in CC units per grab (slow drags accumulate).
    private let latchSweep: Float = 64

    private struct Grab {
        let widget: Widget
        let originY: CGFloat
        var lastShaped: Float   // clamped joystick position at the last event
    }
    private var grabs: [ObjectIdentifier: Grab] = [:]

    // Display mirrors (the engine on the MIDI queue is the truth).
    private var mirrorPitch: Float = 0          // [-1, 1]
    private var mirrorLatch: [Float] = [0, 0, 0]
    private var mirrorSustain = false
    private var mirrorToggleMode = false
    private var mirrorCCs: [UInt8] = [1, 23, 24]
    private var ramping = false

    override init(frame: CGRect) {
        super.init(frame: frame)
        isMultipleTouchEnabled = true
        isOpaque = false
        // Compact floating palette (§8 rev, DECISIONS_3 #31): translucent
        // enough that the marbling reads through, opaque enough to find.
        backgroundColor = UIColor.white.withAlphaComponent(0.42)
        layer.borderWidth = 0.5
        layer.borderColor = UIColor.black.withAlphaComponent(0.15).cgColor
        layer.cornerRadius = 12
        let lp = UILongPressGestureRecognizer(target: self, action: #selector(onLongPress))
        lp.minimumPressDuration = 0.5
        // The editor gesture may only ever see ASSIGNABLE-wheel touches: a
        // recognizer that fires cancels the view's touches (UIKit default),
        // which was releasing a HELD sustain pedal after 0.5 s — the
        // "unpresses while I'm pressing it" device bug (DECISIONS_3 #31).
        lp.delegate = self
        addGestureRecognizer(lp)
    }
    required init?(coder: NSCoder) { fatalError("not used") }

    func gestureRecognizer(_ g: UIGestureRecognizer,
                           shouldReceive touch: UITouch) -> Bool {
        let w = widget(at: touch.location(in: self))
        return w == .assignA || w == .assignB
    }

    // Called by the host whenever engine state changed outside a local touch
    // (mode switches, assignment results, the return ramp finishing).
    func syncMirrors(pitch: Float, latch: [Float], sustain: Bool,
                     toggleMode: Bool, ccs: [UInt8]) {
        mirrorPitch = pitch
        mirrorLatch = latch
        mirrorSustain = sustain
        mirrorToggleMode = toggleMode
        mirrorCCs = ccs
        setNeedsDisplay()
    }

    // -- geometry --------------------------------------------------------------

    private func slotRect(_ w: Widget) -> CGRect {
        let n = CGFloat(Widget.allCases.count)
        let sw = bounds.width / n
        return CGRect(x: CGFloat(w.rawValue) * sw, y: 0, width: sw,
                      height: bounds.height).insetBy(dx: 6, dy: 6)
    }

    private func widget(at p: CGPoint) -> Widget? {
        for w in Widget.allCases where slotRect(w).insetBy(dx: -6, dy: -6).contains(p) {
            return w
        }
        return nil
    }

    // -- touches: the joystick primitive per widget ----------------------------

    // Clamped, knee-shaped vertical deflection for a wheel grab (up = +1).
    private func shaped(_ dyPt: CGFloat) -> Float {
        var ex: Float = 0, ey: Float = 0
        hostmpe_joystick_eff(0, Float(dyPt), wheelTravelPt, &ex, &ey)
        return -ey   // screen y grows down; up is positive
    }

    override func touchesBegan(_ ts: Set<UITouch>, with event: UIEvent?) {
        for t in ts {
            let p = t.location(in: self)
            guard let w = widget(at: p) else { continue }
            // Every widget touch is TRACKED (sustain included): the release
            // must find its widget by the grab record, never by where the
            // finger happens to be when it lifts.
            grabs[ObjectIdentifier(t)] = Grab(widget: w, originY: p.y, lastShaped: 0)
            switch w {
            case .sustain:
                mirrorSustain = mirrorToggleMode ? !mirrorSustain : true
                host?.stripSustainDown()
            case .pitch:
                ramping = false
            case .mod, .assignA, .assignB:
                break
            }
        }
        setNeedsDisplay()
    }

    override func touchesMoved(_ ts: Set<UITouch>, with event: UIEvent?) {
        for t in ts {
            guard var g = grabs[ObjectIdentifier(t)] else { continue }
            let s = shaped(t.location(in: self).y - g.originY)
            switch g.widget {
            case .pitch:
                mirrorPitch = s
                host?.stripPitchMove(s)
            case .mod, .assignA, .assignB:
                // Relative accumulation (§8): only the CHANGE in shaped
                // position moves the value — a regrasp contributes zero.
                let wheel = Int32(g.widget.rawValue - Widget.mod.rawValue)
                let delta = (s - g.lastShaped) * latchSweep
                let idx = Int(wheel)
                mirrorLatch[idx] = min(max(mirrorLatch[idx] + delta, 0), 127)
                host?.stripLatchMove(wheel: wheel, delta: delta)
            case .sustain:
                break
            }
            g.lastShaped = s
            grabs[ObjectIdentifier(t)] = g
        }
        setNeedsDisplay()
    }

    override func touchesEnded(_ ts: Set<UITouch>, with event: UIEvent?) {
        for t in ts {
            guard let g = grabs.removeValue(forKey: ObjectIdentifier(t)) else { continue }
            switch g.widget {
            case .pitch:
                host?.stripPitchRelease()
                animateSpringReturn()
            case .sustain:
                if !mirrorToggleMode { mirrorSustain = false }
                host?.stripSustainUp()   // toggle mode: engine-side no-op
            case .mod, .assignA, .assignB:
                break
            }
        }
        setNeedsDisplay()
    }
    override func touchesCancelled(_ ts: Set<UITouch>, with event: UIEvent?) {
        touchesEnded(ts, with: event)
    }

    // Visual mirror of the engine's 50 ms return ramp (the MIDI ramp itself
    // is emitted by hostmpe_strip_tick on the host's frame drain).
    private func animateSpringReturn() {
        ramping = true
        let v0 = mirrorPitch
        let t0 = CACurrentMediaTime()
        func step() {
            guard ramping else { return }
            let f = Float((CACurrentMediaTime() - t0) / 0.050)
            if f >= 1 {
                mirrorPitch = 0
                ramping = false
            } else {
                mirrorPitch = v0 * (1 - f)
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.016) { step() }
            }
            setNeedsDisplay()
        }
        step()
    }

    // -- long-press: the assignable-wheel CC editor (§8) -----------------------

    @objc private func onLongPress(_ g: UILongPressGestureRecognizer) {
        guard g.state == .began, let w = widget(at: g.location(in: self)),
              w == .assignA || w == .assignB else { return }
        let wheel = Int32(w.rawValue - Widget.mod.rawValue)
        let current = mirrorCCs[Int(wheel)]
        let alert = UIAlertController(
            title: "Assign wheel \(w == .assignA ? "A" : "B")",
            message: "7-bit CC number (0–127). Currently CC \(current). "
                + "Protocol CCs (1, 6, 38, 64, 98–101, 120–127) are refused.",
            preferredStyle: .alert)
        alert.addTextField { tf in
            tf.keyboardType = .numberPad
            tf.text = "\(current)"
        }
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        alert.addAction(UIAlertAction(title: "Assign", style: .default) { [weak self] _ in
            guard let self, let text = alert.textFields?.first?.text,
                  let cc = UInt8(text) else { return }
            self.host?.stripAssign(wheel: wheel, cc: cc) { ok, assigned in
                DispatchQueue.main.async {
                    if ok { self.mirrorCCs[Int(wheel)] = assigned }
                    self.setNeedsDisplay()
                }
            }
        })
        window?.rootViewController?.present(alert, animated: true)
    }

    // -- drawing ---------------------------------------------------------------

    override func draw(_ rect: CGRect) {
        guard let ctx = UIGraphicsGetCurrentContext() else { return }
        let ink = UIColor.black
        for w in Widget.allCases {
            let r = slotRect(w)
            let path = UIBezierPath(roundedRect: r, cornerRadius: 8)
            ctx.setStrokeColor(ink.withAlphaComponent(0.25).cgColor)
            ctx.setLineWidth(1)
            ctx.addPath(path.cgPath)
            ctx.strokePath()

            let label: String
            switch w {
            case .pitch:   label = "Pitch"
            case .mod:     label = "Mod"
            case .assignA: label = "CC \(mirrorCCs[1])"
            case .assignB: label = "CC \(mirrorCCs[2])"
            case .sustain: label = mirrorToggleMode ? "Sus ⇥" : "Sus"
            }
            draw(text: label, in: CGRect(x: r.minX, y: r.maxY - 16,
                                         width: r.width, height: 14), ink: ink)

            if w == .sustain {
                let pad = r.insetBy(dx: r.width * 0.22, dy: r.height * 0.24)
                    .offsetBy(dx: 0, dy: -6)
                let bp = UIBezierPath(roundedRect: pad, cornerRadius: 6)
                if mirrorSustain {
                    ctx.setFillColor(ink.withAlphaComponent(0.45).cgColor)
                    ctx.addPath(bp.cgPath)
                    ctx.fillPath()
                } else {
                    ctx.setStrokeColor(ink.withAlphaComponent(0.4).cgColor)
                    ctx.addPath(bp.cgPath)
                    ctx.strokePath()
                }
                continue
            }

            // Wheel track + thumb.
            let track = CGRect(x: r.midX - 1.5, y: r.minY + 8,
                               width: 3, height: r.height - 32)
            ctx.setFillColor(ink.withAlphaComponent(0.12).cgColor)
            ctx.fill(track)
            let frac: CGFloat   // 0 bottom .. 1 top of the track
            if w == .pitch {
                frac = CGFloat(0.5 + 0.5 * mirrorPitch)
                // center notch
                ctx.setFillColor(ink.withAlphaComponent(0.3).cgColor)
                ctx.fill(CGRect(x: r.midX - 8, y: track.midY - 0.5, width: 16, height: 1))
            } else {
                frac = CGFloat(mirrorLatch[w.rawValue - Widget.mod.rawValue] / 127.0)
            }
            let ty = track.maxY - frac * track.height
            ctx.setFillColor(ink.withAlphaComponent(0.55).cgColor)
            ctx.fillEllipse(in: CGRect(x: r.midX - 7, y: ty - 7, width: 14, height: 14))
        }
    }

    private func draw(text: String, in rect: CGRect, ink: UIColor) {
        let style = NSMutableParagraphStyle()
        style.alignment = .center
        (text as NSString).draw(
            in: rect,
            withAttributes: [.font: UIFont.systemFont(ofSize: 11, weight: .medium),
                             .foregroundColor: ink.withAlphaComponent(0.6),
                             .paragraphStyle: style])
    }
}
